### PR TITLE
[get, subset] chore: skip existing main

### PR DESCRIPTION
**Pull request by uriii3**
Add a skip-existing file option for get and subset.

Solving [CMT-115](https://cms-change.atlassian.net/browse/CMT-115).

- [x] Add option for subset
- [x] Add option for get
- [x] Define possible interactions with overwrite-output-data and sync options
- [x] Add tests to see the behaviour
- [x] Restructure tests if needed
- [x] Add documentation

### CODE REVIEW

**Comment by uriii3** [comment]
For the subset, define the interaction between `skip_existing` and `overwrite_output_data`:
- Currently, the interaction is defined to prioritise the `overwrite_output_data`. That way, if this option is set to true, the code will overwrite the files that are the same, independently if the option `skip_existing` is set to True or not.
- If `dry_run` is set, no data will be downloaded for sure. Still, it will check if the file exists or not, showing a different message in those cases.
- If no `dry_run` and no `overwrite_output_data` but `skip_existing` is set to True, it will not download the file if it is already in the output directory specified.

I added a file within the SubsetResponse, `download_status`. This will display a different message, defined in `get_message_data_downloaded(...)`, regarding the status of the data downloaded.

**Comment by uriii3** [comment]
For the `get` command, I think I've managed to make the option work with the `overwrite_output_data` and the `sync` options, both contemplating the download from direct download as for gathering the files from the bucket:
- As in the subset, the `overwrite_output_data` is prioritised: to make it easy, I throw a warning  if I detect both options and I set back `skip_existing` to False (they are not mutually exclusive, you can set them both, but it will do this).
- For the `sync` option, I thought that for someone doing sync, the added value of the `skip_existing` would be to not even check those that need to be synced. Makes up for a little weird interaction, as of then the option `sync` makes no sense either to have it set to True or not. So I also simplified the work by setting it to False if I detect `skip_existing` and throw a warning.
- With the `dry_run` option, the interaction is also weird. Because I followed the implementation of the `sync` option, actually the `dry_run` only shows the filenames that will be downloaded. In that case, with the `skip_existing` I followed the same behaviour: showing in the `dry_run` the actual relevant information about the download.

File: copernicusmarine/download_functions/download_original_files.py [hunk 1]
```diff
@@ -407,6 +427,14 @@ def _download_header_for_direct_download(
     )
 
 
+def _check_already_exists(
+    filename: str,
+    directory_out: pathlib.Path,
+) -> bool:
+    filename_out = _local_path_from_s3_url(filename, directory_out)
+    return filename_out.is_file()
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_original_files.py:435]: you need to take into account the variable: `no_directories`

File: copernicusmarine/download_functions/download_original_files.py [hunk 2]
```diff
@@ -562,6 +590,7 @@ def _download_one_file(
 def create_filenames_out(
     filenames_in: list[str],
     overwrite: bool,
+    skip_existing: bool,
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_original_files.py:593]: you never use it right?

File: copernicusmarine/download_functions/utils.py [hunk 3]
```diff
@@ -231,10 +231,27 @@ def get_message_formatted_dataset_size_estimation(
...
+) -> str:
+    if dry_run and exists:
+        return "Dry-run, data has not been downloaded (already exists!)."
+    if dry_run:
+        return "Dry-run, data has not been downloaded."
+    if overwrite_output_data:
+        return "Data will be downloaded, with option overwrite_output_data."
+    if exists and skip_existing:
+        return "Data already exists, skipping download."
+    return "Data will be downloaded."
```
Comment (renaudjester) [inline copernicusmarine/download_functions/utils.py:252]: not needed if you do some enums

File: copernicusmarine/core_functions/get.py [hunk 4]
```diff
@@ -85,6 +86,16 @@ def get_function(
         get_request.force_download = force_download
     if overwrite_output_data:
         get_request.overwrite_output_data = overwrite_output_data
+    if overwrite_output_data and skip_existing:
+        logger.warning(
+            "Overwrite output data option is set. Skip existing is ignored."
+        )
+        skip_existing = False
```
[the remark this answers is not in the document]
Comment (uriii3) [reply copernicusmarine/core_functions/get.py:93]: Okey, but: I only see the mutually exclusive option (for `sync` and `no-directories`) with the cli, not with the python interface. Is it there and I'm not finding it or is it just not there?

File: copernicusmarine/download_functions/download_original_files.py (hunk 2 above)
Comment (uriii3) [reply copernicusmarine/download_functions/download_original_files.py:593]: nope, it stayed after a refactoring. 😬

File: copernicusmarine/core_functions/get.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (renaudjester) [reply copernicusmarine/core_functions/get.py:93]: Yeah I think it's missing :O

[the remark this answers is not in the document]
Comment (renaudjester) [reply copernicusmarine/core_functions/get.py:93]: Can you add it please?

[the remark this answers is not in the document]
Comment (renaudjester) [reply copernicusmarine/core_functions/get.py:93]: Or maybe it's just that whatever you set, sync will overwrtie

File: copernicusmarine/core_functions/models.py [hunk 5]
```diff
@@ -85,6 +85,8 @@ class ResponseSubset(BaseModel):
 
     #: Path to the result file.
     output: pathlib.Path
+    #: Message to indicate if the data was downloaded or not.
+    download_status: str
```
[the remark this answers is not in the document]
Comment (uriii3) [reply copernicusmarine/core_functions/models.py:210]: I joined the options to make it easier, made more sense to me. I will do the get now! 😬

File: copernicusmarine/download_functions/download_original_files.py (hunk 1 above)
Comment (uriii3) [reply copernicusmarine/download_functions/download_original_files.py]: Do you think we can make it mutually exclusive with that option too? The way it is built with the sync makes everything make more sense like that, no? (an way easier to code)

File: copernicusmarine/core_functions/subset.py [hunk 6]
```diff
@@ -143,7 +147,13 @@ def subset_function(
     if force_download:
         subset_request.force_download = force_download
     if overwrite_output_data:
+        if skip_existing:
+            raise MutuallyExclusiveArguments(
+                "overwrite_output_data", "skip_existing"
+            )
```
Comment (renaudjester) [inline copernicusmarine/core_functions/subset.py:153]: I think it would be better to have this directly in the CLI and python interface so that the names of the options are consistent

File: copernicusmarine/download_functions/download_arco_series.py [hunk 7]
```diff
@@ -147,6 +152,10 @@ def download_dataset(
         response.status = StatusCode.DRY_RUN
         response.message = StatusMessage.DRY_RUN
         return response
+    elif skip_existing and os.path.exists(output_path):
+        response.status = StatusCode.SUCCESS
+        response.message = StatusMessage.NO_DATA_DOWNLOADED
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_arco_series.py:157]: Statuses cannot be mixed, you have to put the same! so maybe create a status code `SKIP_EXISTING`

File: copernicusmarine/core_functions/models.py [hunk 8]
```diff
@@ -87,6 +92,84 @@ class ResponseGet(BaseModel):
     message: StatusMessage
 
 
+class MiddleResponseGet:
+    """
+    Internal class to have it clear what is
+    returned in the middle of the get function.
```
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:98]: "middle of the function" is not very clear and doesn't seem like something explicit on the long run

File: copernicusmarine/core_functions/models.py [hunk 9]
```diff
@@ -87,6 +92,84 @@ class ResponseGet(BaseModel):
...
+
+    It consists of several lists of the same length,
+    correlating the information in the position of the list.
```
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:101]: Use sub objects

File: copernicusmarine/core_functions/models.py [hunk 10]
```diff
@@ -48,7 +49,11 @@ class StatusMessage(str, Enum):
     DRY_RUN = "The request was successful but no data was transferred."
     ERROR = "An error occurred during the request."
     NO_DATA_TO_DOWNLOAD = "No data to download from the remote server corresponding to your request."  # noqa: E501
+    NO_DATA_DOWNLOADED = "No data was downloaded."
+    FILES_IGNORED = "The request was successful, and it ignored some files."
+    FILES_DELETED = "The request was successful, and it deleted some files."
     FILE_LIST_CREATED = "The request created a file list and then stopped."
+    OVERWRITED_DATA = "The request overwrited the data."
```
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:56]: StatusCode and StatusMessage were suppose to be 1:1

File: copernicusmarine/core_functions/get.py [hunk 11]
```diff
@@ -77,14 +81,38 @@ def get_function(
...
+        if skip_existing:
+            raise MutuallyExclusiveArguments(
+                "overwrite_output_data", "skip_existing"
+            )
+        elif sync:
+            raise MutuallyExclusiveArguments("overwrite_output_data", "sync")
+        elif sync_delete:
+            raise MutuallyExclusiveArguments(
+                "overwrite_output_data", "sync_delete"
+            )
```
Comment (renaudjester) [inline copernicusmarine/core_functions/get.py:99]: Same as before, do it on the CLI and python interface level

File: copernicusmarine/core_functions/exceptions.py [hunk 12]
```diff
@@ -90,3 +90,16 @@ class ServiceNotSupported(Exception):
 
     def __init__(self, service_type):
         super().__init__(f"Service type {service_type} not supported.")
+
+
+class MutuallyExclusiveArguments(Exception):
+    """
+    Exception raised when mutually exclusive arguments are used together.
+
+    Please make sure the arguments are not used together.
```
Comment (renaudjester) [inline copernicusmarine/core_functions/exceptions.py:99]: Maybe not necessary

File: Makefile [hunk 13]
```diff
@@ -60,7 +60,7 @@ run-using-pyinstaller-windows-latest:
 
 run-using-pyinstaller-macos:
 	pip install -e .
-	python -m PyInstaller --collect-all tzdata --copy-metadata copernicusmarine --name copernicusmarine_macos-${ARCH}.cli --icon=toolbox_icon.png copernicusmarine/command_line_interface/copernicus_marine.py --onefile --target-architecture=${ARCH}
+	python -m PyInstaller --collect-all tzdata --copy-metadata copernicusmarine --name copernicusmarine_macos-${ARCH}.cli --icon=toolbox_icon.png --add-data="/home/runner/micromamba/envs/copernicusmarine-binary/lib/python3.12/site-packages/distributed/distributed.yaml:./distributed" copernicusmarine/command_line_interface/copernicus_marine.py --onefile --target-architecture=${ARCH}
```
Comment (renaudjester) [inline Makefile:63]: For another PR?

File: copernicusmarine/download_functions/download_original_files.py [hunk 14]
```diff
@@ -175,14 +160,18 @@ def download_original_files(
...
-    bucket: str
-    endpoint, bucket = locator
 
+    endpoint, bucket = middleResponse.locator
+
+    message_output = StatusMessage.SUCCESS
+    if len(middleResponse.filenames_to_be_ignored) > 0:
+        message_output = StatusMessage.FILES_IGNORED
     if get_request.sync_delete and files_to_delete:
+        message_output = StatusMessage.FILES_DELETED
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_original_files.py:170]: I think you can create a message per file

Comment (uriii3) [reply copernicusmarine/download_functions/download_original_files.py:170]: `sync_delete` and `skip_existing` are mutually exclusive, so it wouldn't need to happen. And also, the class constructed is not a full list for all the objects, maybe we could think of it that way, but the structure was thought to keep what you did... however you want to treat it!

File: Makefile (hunk 13 above)
Comment (uriii3) [reply Makefile:63]: Yep, I created another one to be able to check.

File: copernicusmarine/core_functions/subset.py (hunk 6 above)
Comment (uriii3) [reply copernicusmarine/core_functions/subset.py:153]: Done!

File: copernicusmarine/download_functions/download_arco_series.py (hunk 7 above)
Comment (uriii3) [reply copernicusmarine/download_functions/download_arco_series.py:157]: Hehe I added some numbers to group them in success or not success, so maybe it can be interesting, but I think you won't like it and it will go down again hehe

File: copernicusmarine/core_functions/exceptions.py (hunk 12 above)
Comment (uriii3) [reply copernicusmarine/core_functions/exceptions.py:99]: I like it, but however you want to keep it.

File: copernicusmarine/core_functions/utils.py [hunk 15]
```diff
@@ -44,20 +44,19 @@
 
 
 def get_unique_filename(
-    filepath: pathlib.Path, overwrite_option: bool
+    filepath: pathlib.Path, overwrite_option: bool, skip_existing: bool
```
Comment (renaudjester) [inline copernicusmarine/core_functions/utils.py:47]: do we need it since the file should just be ignored right? we shouldn't even need to create the unique file name

File: tests/__snapshots__/test_help_command_interface.ambr [hunk 16]
```diff
@@ -117,7 +118,9 @@
     '                                  If specified and if the file already exists',
     '                                  on destination, then it will be overwritten',
     '                                  instead of creating new one with unique',
-    '                                  index.',
+    '                                  index.                 \\x08 NOTE: This argument',
+    '                                  is mutually exclusive with arguments:',
+    '                                  [skip_existing, sync, sync_delete].',
```
Comment (renaudjester) [inline tests/__snapshots__/test_help_command_interface.ambr:123]: does it look good in the terminal?

File: tests/test_command_line_interface.py [hunk 17]
```diff
@@ -1860,3 +1863,56 @@ def test_coordinates_selection_method_outside_w_elevation(self, tmp_path):
...
+
+    def test_all_mutually_exclusive_arguments(self):
+        self.when_mutually_exclusive_arguments(
+            "get", "--skip-existing", "--overwrite-outputdata"
+        )
+        self.when_mutually_exclusive_arguments(
+            "get", "--skip-existing", "--sync"
+        )
+        self.when_mutually_exclusive_arguments(
+            "get", "--skip-existing", "--sync-delete"
```
Comment (renaudjester) [inline tests/test_command_line_interface.py:1875]: Can you put the mutually exclusive tests in an other file?

File: copernicusmarine/core_functions/documentation_utils.py [hunk 18]
```diff
@@ -49,6 +49,11 @@
         "please refer to the documentation or use option ``--create-template`` "
         "from the command line interface for an example template."
     ),
+    "SKIP_EXISTING_HELP": (
+        "If the files already exists where it would be downloaded, then "
+        "the download is skipped for this file. By default, the toolbox "
+        "creates a new file with an index (eg 'filename_(1).nc'."
```
[the remark this answers is not in the document]
Comment (uriii3) [reply copernicusmarine/core_functions/documentation_utils.py]: It is also true for the get!

File: copernicusmarine/core_functions/models.py [hunk 19]
```diff
@@ -87,6 +100,73 @@ class ResponseGet(BaseModel):
...
+    #     self.total_size = 0.0
+    #     self.filenames_to_be_ignored = []
+    #     self.files_not_found = []
+    #     self.is_None = False
+
+    def add(self, S3Fileinformation):
+        self.s3_file_list.append(S3Fileinformation)
+
+    def append_more_info(self, middle_response):
+        self.locator = middle_response.locator
```
[the remark this answers is not in the document]
Comment (uriii3) [reply copernicusmarine/core_functions/models.py:155]: I deleted it hehe

File: copernicusmarine/core_functions/models.py
[the remark this answers is not in the document]
Comment (uriii3) [reply copernicusmarine/core_functions/models.py:1]: I created a new type of class for the files that we will not get (that way, we don't need to worry about all the information that needs to be get).

That way, I can add some of the files into the 'list', with the correct status.

File: copernicusmarine/download_functions/download_original_files.py [hunk 20]
```diff
@@ -175,38 +157,40 @@ def download_original_files(
...
                 file_format=filename_out.suffix,
             )
-            for s3_url, size, last_modified, filename_out, etag in zip(
-                filenames_in,
-                sizes,
-                last_modified_datetimes,
-                filenames_out,
-                etags,
+            for s3_file, filename_out in zip(
+                middleResponse.s3_file_list, filenames_out
```
[the remark this answers is not in the document]
Comment (uriii3) [reply copernicusmarine/download_functions/download_original_files.py:191]: Mmh I think it already works like this, maybe I prefer it this way (so I can have the typings fixed and not add the filenames_out a posteriori)

File: copernicusmarine/download_functions/download_original_files.py [hunk 21]
```diff
@@ -420,36 +381,45 @@ def _download_header_for_direct_download(
...
+                    middle_response.add(
+                        S3FileInfo(
+                            filename=full_path,
+                            size=float(size),
+                            last_modified_datetime=last_modified,
+                            etag=etag,
+                        )
+                    )
+                else:
+                    middle_response.files_to_be_ignored.append(full_path)
```
[the remark this answers is not in the document]
Comment (uriii3) [reply copernicusmarine/download_functions/download_original_files.py:402]: Yes, you are right!

File: tests/__snapshots__/test_help_command_interface.ambr (hunk 16 above)
Comment (uriii3) [reply tests/__snapshots__/test_help_command_interface.ambr]: Yes it does!

Comment (uriii3) [reply tests/__snapshots__/test_help_command_interface.ambr]: I mean it adds some spaces into the description, I'll check if I can create a new line and maybe it will look better.

Comment (uriii3) [reply tests/__snapshots__/test_help_command_interface.ambr]: I removed the \\x08, as it was not properly creating a new line anyway.

File: tests/test_command_line_interface.py (hunk 17 above)
Comment (uriii3) [reply tests/test_command_line_interface.py]: Sure!

File: conda_environment_test.yaml [hunk 22]
```diff
@@ -11,3 +11,4 @@ dependencies:
   - pip:
     - pytest-order==1.2.1
     - freezegun==1.5.1
+    - pydantic-extra-types==2.10.0
```
Comment (renaudjester) [inline conda_environment_test.yaml:14]: why is this needed?

File: copernicusmarine/command_line_interface/group_get.py [hunk 23]
```diff
@@ -153,15 +155,23 @@ def cli_get() -> None:
     is_flag=True,
     default=False,
     help=documentation_utils.GET["SYNC_HELP"],
-    mutually_exclusive=["no-directories"],
+    mutually_exclusive=["no-directories", "skip_existing"],
```
Comment (renaudjester) [inline copernicusmarine/command_line_interface/group_get.py:158]:
```suggestion
    mutually_exclusive=["no-directories", "skip-existing"],
```

File: copernicusmarine/command_line_interface/group_get.py [hunk 24]
```diff
@@ -153,15 +155,23 @@ def cli_get() -> None:
...
 )
 @click.option(
     "--sync-delete",
     cls=MutuallyExclusiveOption,
     is_flag=True,
     default=False,
     help=documentation_utils.GET["SYNC_DELETE_HELP"],
-    mutually_exclusive=["no-directories"],
+    mutually_exclusive=["no-directories", "skip_existing"],
```
Comment (renaudjester) [inline copernicusmarine/command_line_interface/group_get.py:166]:
```suggestion
    mutually_exclusive=["no-directories", "skip-existing"],
```

File: copernicusmarine/command_line_interface/group_get.py [hunk 25]
```diff
@@ -77,7 +77,7 @@ def cli_get() -> None:
     is_flag=True,
     help=documentation_utils.GET["NO_DIRECTORIES_HELP"],
     default=False,
-    mutually_exclusive=["sync"],
+    mutually_exclusive=["sync", "sync_delete", "skip_existing"],
```
Comment (renaudjester) [inline copernicusmarine/command_line_interface/group_get.py:80]:
```suggestion
    mutually_exclusive=["sync", "sync-delete", "skip-existing"],
```

File: copernicusmarine/command_line_interface/group_get.py [hunk 26]
```diff
@@ -107,7 +107,9 @@ def cli_get() -> None:
     documentation_utils.GET["OVERWRITE_SHORT_OPTION"],
     is_flag=True,
     default=False,
+    cls=MutuallyExclusiveOption,
     help=documentation_utils.GET["OVERWRITE_OUTPUT_DATA_HELP"],
+    mutually_exclusive=["skip_existing", "sync", "sync_delete"],
```
Comment (renaudjester) [inline copernicusmarine/command_line_interface/group_get.py:112]:
```suggestion
    mutually_exclusive=["skip-existing", "sync", "sync-delete"],
```

File: copernicusmarine/command_line_interface/group_get.py [hunk 27]
```diff
@@ -153,15 +155,23 @@ def cli_get() -> None:
...
-    mutually_exclusive=["no-directories"],
+    mutually_exclusive=["no-directories", "skip_existing"],
+)
+@click.option(
+    "--skip-existing",
+    cls=MutuallyExclusiveOption,
+    is_flag=True,
+    default=False,
+    help=documentation_utils.GET["SKIP_EXISTING_HELP"],
+    mutually_exclusive=["no-directories", "sync", "sync_delete"],
```
Comment (renaudjester) [inline copernicusmarine/command_line_interface/group_get.py:174]:
```suggestion
    mutually_exclusive=["no-directories", "sync", "sync-delete"],
```

File: copernicusmarine/command_line_interface/utils.py [hunk 28]
```diff
@@ -10,11 +10,13 @@ def __init__(self, *args, **kwargs):
         self.mutually_exclusive = set(kwargs.pop("mutually_exclusive", []))
         help = kwargs.get("help", "")
         if self.mutually_exclusive:
-            ex_str = ", ".join(self.mutually_exclusive)
+            list_mutually_exclusive = list(self.mutually_exclusive)
+            list_mutually_exclusive.sort()
+            ex_str = ", ".join(list_mutually_exclusive)
```
Comment (renaudjester) [inline copernicusmarine/command_line_interface/utils.py:15]:
```suggestion
            ex_str = ", ".join(sorted(self.mutually_exclusive))
```

File: copernicusmarine/core_functions/models.py [hunk 29]
```diff
@@ -46,12 +49,24 @@ class StatusMessage(str, Enum):
 
     SUCCESS = "The request was successful."
     DRY_RUN = "The request was successful but no data was transferred."
-    ERROR = "An error occurred during the request."
-    NO_DATA_TO_DOWNLOAD = "No data to download from the remote server corresponding to your request."  # noqa: E501
     FILE_LIST_CREATED = "The request created a file list and then stopped."
+    NO_DATA_DOWNLOADED = "No data was downloaded."
```
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:53]: Maybe this one is not necessary

File: copernicusmarine/core_functions/models.py [hunk 30]
```diff
@@ -46,12 +49,24 @@ class StatusMessage(str, Enum):
...
+    ERROR = "An error occurred during the request."
+
+
+class StatusFile(str, Enum):
+    """
+    Status of the files in the get
+    """
+
+    DOWNLOAD = "DOWNLOAD"
+    IGNORED = "IGNORED"
```
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:65]: Use it also for the subset while you have it

Comment (renaudjester) [reply copernicusmarine/core_functions/models.py:65]: Also because in the future we might have multiple files for the subset the response so it might work in a similar fashion

File: copernicusmarine/core_functions/models.py [hunk 31]
```diff
@@ -87,6 +108,44 @@ class ResponseGet(BaseModel):
...
+    # List of information of files that will be downloaded
+    s3_file_list: list[S3FileInfo] = []
+    # Total size (not sure yet if the total donwload or all the files)
+    total_size: float = 0.0
+    # List of all the files
+    # (the ones to be ignored and the ones that will be downloaded)
+    files_all_files: list[str] = []
+    files_ignored: list[S3FileInfo] = []
+    files_not_found: list[str] = []
+    # A bool to know if creates a file list. If is None, it is True.
```
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:145]: I would delete the comments for clarity and simplicity

File: copernicusmarine/core_functions/models.py [hunk 32]
```diff
@@ -87,6 +108,44 @@ class ResponseGet(BaseModel):
...
+    model_config = ConfigDict(use_enum_values=True)
+    locator: tuple[str, str] = ("", "")
+    # List of information of files that will be downloaded
+    s3_file_list: list[S3FileInfo] = []
+    # Total size (not sure yet if the total donwload or all the files)
+    total_size: float = 0.0
+    # List of all the files
+    # (the ones to be ignored and the ones that will be downloaded)
+    files_all_files: list[str] = []
+    files_ignored: list[S3FileInfo] = []
```
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:143]: Put this info in the `S3FileInfo`

File: copernicusmarine/core_functions/models.py (hunk 30 above)
Comment (renaudjester) [inline copernicusmarine/core_functions/models.py:65]:
```suggestion  
    DOWNLOADED = "DOWNLOADED"
    IGNORED = "IGNORED"
    OVERWRITTEN = "OVERWRITTEN"
```

And then add descriptions and add this to the documentation by putting this class in the `__init__` and the reponse type rst file

File: copernicusmarine/core_functions/utils.py [hunk 33]
```diff
@@ -31,20 +31,20 @@
 
 
 def get_unique_filename(
-    filepath: pathlib.Path, overwrite_option: bool
+    filepath: pathlib.Path,
+    same_filename: bool,
```
Comment (renaudjester) [inline copernicusmarine/core_functions/utils.py:35]: I don't like this name 🤔 

what about: `create_new_file`: true if you want a new file (filename_(1)) and false otherwise
other proposition `add_counter_if_needed`

File: copernicusmarine/download_functions/download_arco_series.py [hunk 34]
```diff
@@ -147,6 +151,10 @@ def download_dataset(
         response.status = StatusCode.DRY_RUN
         response.message = StatusMessage.DRY_RUN
         return response
+    elif skip_existing and os.path.exists(output_path):
+        response.status = StatusCode.NO_DATA_DOWNLOADED
+        response.message = StatusMessage.NO_DATA_DOWNLOADED
+        return response
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_arco_series.py:157]: I would keep this as success but add a line in the subset response (like for the FileGet) that says the status of the file

File: copernicusmarine/download_functions/download_arco_series.py [hunk 35]
```diff
@@ -157,6 +165,9 @@ def download_dataset(
     )
     download_delayed_dataset(delayed, disable_progress_bar)
     logger.info(f"Successfully downloaded to {output_path}")
+    if overwrite_output_data:
+        response.status = StatusCode.DATA_OVERWRITTEN
+        response.message = StatusMessage.DATA_OVERWRITTEN
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_arco_series.py:170]: Same here

File: copernicusmarine/core_functions/models.py (hunk 29 above)
Comment (renaudjester) [reply copernicusmarine/core_functions/models.py:53]: The `DATA_OVERWRITTEN` neither by the way

File: copernicusmarine/download_functions/download_original_files.py [hunk 36]
```diff
@@ -296,26 +342,14 @@ def _download_header(
...
     (endpoint_url, bucket, path) = parse_access_dataset_url(
         data_path, only_dataset_root_path=only_list_root_path
     )
 
-    filenames: list[str] = []
-    sizes: list[float] = []
-    total_size = 0.0
-    last_modified_datetimes: list[DateTime] = []
-    etags: list[str] = []
+    middle_response = S3FilesDescriptor(locator=(endpoint_url, bucket))
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_original_files.py:351]: middle response no

File: copernicusmarine/download_functions/download_original_files.py [hunk 37]
```diff
@@ -324,55 +358,74 @@ def _download_header(
...
+                filename, directory_out
             ):
-                filenames.append(filename)
-                sizes.append(float(size))
-                last_modified_datetimes.append(last_modified_datetime)
-                etags.append(etag)
-    total_size = sum(sizes)
+                if not sync or _check_needs_to_be_synced(
+                    filename, size, last_modified_datetime, directory_out
+                ):
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_original_files.py:370]: Do the if's after or combine them but you can do way less lines for this

File: copernicusmarine/download_functions/download_original_files.py [hunk 38]
```diff
@@ -381,35 +434,24 @@ def _download_header_for_direct_download(
...
-    float,
-    list[str],
-    list[str],
-    list[str],
-]:
+    skip_existing: bool,
+) -> S3FilesDescriptor:
     (endpoint_url, bucket, path) = parse_access_dataset_url(dataset_url)
+
+    middle_response = S3FilesDescriptor(locator=(endpoint_url, bucket))
```
Comment (renaudjester) [inline copernicusmarine/download_functions/download_original_files.py:441]: no middle response

File: doc/usage/shared-options.rst [hunk 39]
```diff
@@ -251,3 +251,18 @@ All logs of the library are by default logged in stderr, except the output of th
 
 .. note::
     For versions <=1.2.4, all logs are sent to stdout by default.
+
+
+Option ``--skip-existing``
+*******************************
+
+This option allows you to skip the download of files that already exist in the output directory. If the file exists, the download will be skipped, and the file will not be overwritten.
```
Comment (renaudjester) [inline doc/usage/shared-options.rst:259]:
```suggestion
This option allows you to skip the download of files that already exist at the output destination. If the file exists, the download will be skipped.
```

File: doc/usage/shared-options.rst [hunk 40]
```diff
@@ -251,3 +251,18 @@ All logs of the library are by default logged in stderr, except the output of th
...
+
+Remember that by default, the toolbox creates a new file with an index (eg 'filename_(1).nc').
+
+See in :ref:`Response types documentation <response-types>` the ``status`` and ``message`` fields to understand the status of your request.
```
Comment (renaudjester) [inline doc/usage/shared-options.rst:263]:
```suggestion
See in :ref:`Response types documentation <response-types>` the ``status`` and ``message`` fields for more information about request statuses.
```

File: doc/usage/shared-options.rst [hunk 41]
```diff
@@ -251,3 +251,18 @@ All logs of the library are by default logged in stderr, except the output of th
...
+
+.. note::
+    This option is incompatible in ``get`` command with the ``--sync``, ``--sync-delete``, ``--no-directories`` and ``--overwrite-outputdata`` option.
+
+    For the ``subset`` command, it is incompatible with the ``--overwrite-outputdata`` option.
```
Comment (renaudjester) [inline doc/usage/shared-options.rst:268]:
```suggestion
.. note::
    This option is incompatible with the ``--overwrite-output-data`` option and the ``--sync`` and ``--sync-delete`` options.
```

File: tests/test_command_line_interface.py [hunk 42]
```diff
@@ -319,7 +320,11 @@ def test_downloaded_files_are_not_displayed_with_force_download_option(
             not in self.output.stderr
         )
 
-    def test_get_by_default_returns_status_message(self, tmp_path):
+    def test_something_and_skip_existing(self, tmp_path):
```
Comment (renaudjester) [inline tests/test_command_line_interface.py:323]:
```suggestion
    def test_get_something_and_skip_existing(self, tmp_path):
```

File: copernicusmarine/core_functions/models.py (hunk 30 above)
Comment (uriii3) [reply copernicusmarine/core_functions/models.py:65]: doesn't make sense to use it in the subset, as there is not a list of files there. There is only one file, which already has a status. I don't exactly see the point (we can use it if we do return a list of files with the subset also!)

File: copernicusmarine/core_functions/models.py (hunk 29 above)
Comment (uriii3) [reply copernicusmarine/core_functions/models.py:53]: Okay :)

File: copernicusmarine/core_functions/models.py (hunk 32 above)
Comment (uriii3) [reply copernicusmarine/core_functions/models.py:143]: What info?

File: copernicusmarine/core_functions/utils.py (hunk 33 above)
Comment (uriii3) [reply copernicusmarine/core_functions/utils.py:35]: I kind of like it because it follows better the intuition of the function: if overwritten, same_filename is true. If skip_existing, same filename is true. That way, it is easier to think about the creation of the filenames and more intuitive, not having to think the negation of the boolean constantly.

File: copernicusmarine/download_functions/download_original_files.py (hunk 36 above)
Comment (uriii3) [reply copernicusmarine/download_functions/download_original_files.py:351]: :(

File: copernicusmarine/download_functions/download_original_files.py (hunk 37 above)
Comment (uriii3) [reply copernicusmarine/download_functions/download_original_files.py:370]: I think you refer on the later lines, I've simplified it a bit, let me know if it fits what you were saying!

File: copernicusmarine/core_functions/models.py (hunk 30 above)
Comment (uriii3) [reply copernicusmarine/core_functions/models.py:61]: Done!

File: conda_environment_test.yaml (hunk 22 above)
Comment (uriii3) [reply conda_environment_test.yaml:14]: Okay, yes, it looks like I needed ti to define the DateTime field inside a class. Maybe we need it for the whole toolbox then too, no?
